src/Abp.Zero.EntityFramework/EntityHistory/EntityHistoryHelper.cs

```diff
@@ -487,54 +475,94 @@ private void UpdateChangeSet(DbContext context, EntityChangeSet changeSet)
+                        if (propertyEntry.HasChanged())
+                        {
+                            // Add foreign key
+                            entityChange.PropertyChanges.Add(CreateEntityPropertyChange(oldValue, newValue, propertyInfo));
+                        }
+
+                        continue;
+                    }
 
-                        //make sure test case cover post saving (for foreign key update)
-                        if (propertyChange == null)
+                    if (propertyChange.OriginalValue == propertyChange.NewValue)
+                    {
+                        var newPropertyChange = CreateEntityPropertyChange(oldValue, newValue, propertyInfo);
+                        if (propertyChange.NewValue == newPropertyChange.NewValue)
+                        {
+                            // No change
+                            entityChange.PropertyChanges.Remove(propertyChange);
+                        }
+                        else
                         {
-                            if (propertyEntry.HasChanged())
-                            {
-                                var propertyInfo = entityEntry.GetPropertyInfo(property.Name);
-
-                                // Add foreign key
-                                entityChange.PropertyChanges.Add(new EntityPropertyChange
-                                {
-                                    NewValue = propertyEntry.CurrentValue.ToJsonString(),
-                                    OriginalValue = propertyEntry.OriginalValue.ToJsonString(),
-                                    PropertyName = property.Name,
-                                    PropertyTypeFullName = propertyInfo.PropertyType.FullName
-                                });
-                            }
-
-                            continue;
+                            // Update foreign key
+                            propertyChange.NewValue = newPropertyChange.NewValue;
                         }
+                    }
+                }
+
+                /* Update tracked properties */
+                var auditedPropertyNames = entityChange.PropertyChanges
+                                             .Select(pc => pc.PropertyName)
+                                             .Except(foreignKeyPropertyNames)
+                                             .Distinct()
+                                             .ToList();
 
-                        if (propertyChange.OriginalValue == propertyChange.NewValue)
+                foreach(var propertyName in auditedPropertyNames)
+                {
+                    var memberEntry = entityEntry.Member(propertyName);
+                    var primitiveOrComplexProperty = memberEntry is DbPropertyEntry || memberEntry is DbComplexPropertyEntry;
+                    if (!primitiveOrComplexProperty)
```

Comment:
```diff
- var primitiveOrComplexProperty = memberEntry is DbPropertyEntry || memberEntry is DbComplexPropertyEntry;
- if (!primitiveOrComplexProperty)
+ if (!(memberEntry is DbPropertyEntry))
```

[`DbComplexPropertyEntry`](https://docs.microsoft.com/en-us/dotnet/api/system.data.entity.infrastructure.dbcomplexpropertyentry?view=entity-framework-6.2.0) inherits [`DbPropertyEntry`](https://docs.microsoft.com/en-us/dotnet/api/system.data.entity.infrastructure.dbpropertyentry?view=entity-framework-6.2.0) so `|| memberEntry is DbComplexPropertyEntry` is redundant.

In fact, this check itself is redundant since `propertyName` here is from `entityChange.PropertyChanges`, which already did this check.

Reply:
> DbComplexPropertyEntry inherits DbPropertyEntry so || memberEntry is DbComplexPropertyEntry is redundant.

You're right, checking `DbPropertyEntry` will be sufficient.

> In fact, this check itself is redundant since propertyName here is from entityChange.PropertyChanges, which already did this check.

`entityChange.PropertyChanges` contains relationship changes as well. so we need to filter out relationship properties.

Reply:
I see. That is not obvious (and the naming seems counter-intuitive), so please add a comment.

Reply:
added comments